# HG changeset patch
# User Rik <rik@octave.org>
# Date 1475250678 25200
#      Fri Sep 30 08:51:18 2016 -0700
# Branch stable
# Node ID 3fd1b248bb27bd3d166e02cd79b2722c82a5ae9e
# Parent  b828361c8949b994d80e8dcfb8d6dda4907a54fc
Better Matlab compatibility for images created with gnuplot (bug #49130).

* __gnuplot_draw_axes__.m (mapcdata): Don't do conversion to
double->uint8->double to achieve clamping of RGB values.  Keep cdata as double
and use indexing to find and clamp values outside range [0, 255].
For scaled data, use scaling factor of colormap_size rather than
colormap_size-1 for compatibility.
For "direct" integer data, convert range from zero-based indices to ones-based.
For "direct" float data, truncate using "fix" rather than "round" for
compatibility.

diff --git a/scripts/plot/util/private/__gnuplot_draw_axes__.m b/scripts/plot/util/private/__gnuplot_draw_axes__.m
--- a/scripts/plot/util/private/__gnuplot_draw_axes__.m
+++ b/scripts/plot/util/private/__gnuplot_draw_axes__.m
@@ -2844,26 +2844,32 @@ function maybe_do_x2tick_mirror (plot_st
     fprintf (plot_stream, "unset xtics; set x2tics %s nomirror\n",
                           axis_obj.tickdir);
   endif
 
 endfunction
 
 function retval = mapcdata (cdata, mode, clim, cmap_sz)
   if (ndims (cdata) == 3)
-    # True color, clamp data to 8-bit
+    ## True Color, clamp data to 8-bit
     cdata = double (cdata);
-    cdata = uint8 (255*(cdata-clim(1))/(clim(2)-clim(1)));
-    # Scale using inverse of gnuplot's cbrange mapping
-    retval = 1 + double (cdata)*(cmap_sz-1)/255;
+    cdata = 255 * (cdata - clim(1)) / (clim(2)-clim(1));
+    cdata(cdata < 0) = 0;  cdata(cdata > 255) = 255;
+    ## Scale using inverse of gnuplot's cbrange mapping
+    retval = 1 + cdata * (cmap_sz-1)/255;
   else
     if (islogical (cdata))
       cdata += 1;
     elseif (strcmp (mode, "scaled"))
       cdata = double (cdata);
       clim = double (clim);
-      cdata = 1 + fix ((cmap_sz-1)*(cdata-clim(1))/(clim(2)-clim(1)));
+      cdata = 1 + fix (cmap_sz * (cdata - clim(1)) / (clim(2) - clim(1)));
     else
-      cdata = round (cdata);
+      if (isinteger (cdata))
+        cdata += 1;
+      else
+        cdata = fix (cdata);
+      endif
     endif
     retval = max (1, min (cdata, cmap_sz));
   endif
 endfunction
+
